### PR TITLE
Mass-conditional RNN

### PR DESCRIPTION
**Pull request by vineetbansal**
A WIP based on the work in PR #239, but proceeding in a series of individually-tested commits.

### CODE REVIEW

**Comment by skinnider** [comment]
Made it through the commits below - so far so good, except questions/points noted ... 
[094fa31](https://github.com/skinniderlab/CLM/pull/249/commits/094fa31ddb21eb9cf16f6a6639a054f632dba7ab)
[9cbff85](https://github.com/skinniderlab/CLM/pull/249/commits/9cbff853f01201090fde11047f1bd444cbaf27c1)
[7fb664c](https://github.com/skinniderlab/CLM/pull/249/commits/7fb664c1592764c4d109486a7729c9195e8b1add)
[6a5f60a](https://github.com/skinniderlab/CLM/pull/249/commits/6a5f60a9fb24743a294ab6fe47483f695778dfc1)
[1a65931](https://github.com/skinniderlab/CLM/pull/249/commits/1a65931184bf8a0360e334fd3871a7e0ba943c00)
[ef79b23](https://github.com/skinniderlab/CLM/pull/249/commits/ef79b2355e3bd9e4c100ac85c7cd9855c8277340)
[e1efabe](https://github.com/skinniderlab/CLM/pull/249/commits/e1efabe4bb6dc6d4a9c4239340fb8f7844cd0329)
- still contains code related to difference between original/computed descriptors (`if self.gamma > 0: ...`) - this should probably (?) be removed because in the future our descriptors will be experimentally obtained, not calculable
- parameter names/arguments are inconsistent - `mass_*` (from original code) vs. `conditional_*` (adapted for more than just masses)
[cc67a82](https://github.com/skinniderlab/CLM/pull/249/commits/cc67a8261318496e5bd10f4b76435ec4f1071cca)
[34db6ca](https://github.com/skinniderlab/CLM/pull/249/commits/34db6caeec3f535b75a519162907006ad4eff5bf)
- sampling is being done based on ranges of descriptors in the training set - but should instead be based on the descriptors associated with the held-out validation set
[731cb33](https://github.com/skinniderlab/CLM/pull/249/commits/731cb33fa39ac91bc5aea222986e1eb2bea62ce3)
[4a2282c](https://github.com/skinniderlab/CLM/pull/249/commits/4a2282ce017334aa102a5d292582dcf3feb6376c)
[13cc36c](https://github.com/skinniderlab/CLM/pull/249/commits/13cc36c8df6f8bab3b0c0d1827edb225de8d56d4)
[3c42d69](https://github.com/skinniderlab/CLM/pull/249/commits/3c42d69132a89d4934b754d68199bb9232453b79)
[49cbafc](https://github.com/skinniderlab/CLM/pull/249/commits/49cbafc9b01d9e48d5d7ecf1a6fa4bd3663e056e)

**Recap of conditional RNN parameters (note to self, for future reference):**
- conditional_emb concatenates the descriptors directly to the token embeddings at each step in the sequence; conditional_emb_l first passes them through a linear layer to obtain embeddings of dimensionality equal to that of the token embeddings
- conditional_dec and conditional_dec_l are analogous, except that the descriptors or their embeddings are concatenated to the output of the RNN layers (prior to the decoder layer)
- conditional_h instantiates the hidden states based on learned transformations of the descriptors (with a single linear layer), as in Kotsias et al.

**Comment by skinnider** [comment]
Last two commits:
[3debdc7](https://github.com/skinniderlab/CLM/pull/249/commits/3debdc7a4b049ad686c99be9745e821c54fb8f2c)
[780d1a0](https://github.com/skinniderlab/CLM/pull/249/commits/780d1a09d06c3c7479c35550ab468b7e12af52f4)
... both look OK.

So, I think my only comments are:
1. In [models.py](https://github.com/skinniderlab/CLM/blob/vb/massrnn/src/clm/models.py#L693), there is code that starts on L694 that calculates a loss between the (computed) descriptors of the generated molecules and the descriptors in the training set. I think this would be a great functionality if we could make it work, but I am not convinced this is very generalizable, because it assumes the computed descriptors are always going to be the same across any application of the RNN. In some cases, the descriptors will be derived experimentally, so we can't simply recompute them from the generated SMILES. Also, I think this code references a `SmilesDescriptorCollates.get_descriptors` function that is not imported (?)
2. The nomenclature of parameters and variables switches around quite a bit between `conditional_*`/`mass(es)_*`/`descriptor(s)_*`. It may be clearer to future readers if we can replace all instances of `masses_*` with `descriptors*`, since it's no longer the case that the only descriptor of interest is a mass.

**Comment by vineetbansal** [comment]
Thanks @skinnider - makes sense. I've made these suggested changes in commit 3433dfe on this branch.

**Comment by skinnider** [comment]
Looks great!

**Review by skinnider — APPROVED** [review]
